sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs

```diff
@@ -986,18 +986,19 @@ private async Task<AmqpSession> CreateAndOpenSessionAsync(AmqpConnection connect
             {
                 var openObjectCompletionSource = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
 
-                using var registration = cancellationToken.Register(static state =>
+                using var registration = cancellationToken.Register(state =>
                 {
                     var tcs = (TaskCompletionSource<object>)state;
                     tcs.TrySetCanceled();
+                    target.SafeClose();
                 }, openObjectCompletionSource, useSynchronizationContext: false);
 
                 static async Task Open(AmqpObject target, TimeSpan timeout, TaskCompletionSource<object> openObjectCompletionSource)
                 {
                     try
                     {
                         await target.OpenAsync(timeout).ConfigureAwait(false);
-                        openObjectCompletionSource.TrySetResult(null);
+                        openObjectCompletionSource.SetResult(null);
```

Comment:
I would recommend against this. Depending on order of things it is possible that the task completion source will be tried to set into different states. The idea is that always the first state transition wins and further one's are ignored. With the Set API this would start throwing which is not desirable. That's why I deliberately took Try methods